Reduce the number of variables in the SQL query

Currently the integration test throws the following error:

```
E       sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) too many SQL variables
```

(it goes on for quite some time, but it's referring to the giant CASE clause that is the species lookup)

This is a known limitation with SQLite where one is limited to 999 "`?`" variables. Ref: https://www.sqlite.org/limits.html

Upon further investigation, I see that the insert expressions are going to have issues as well for any large-scale operations (though this can be mitigated by setting `batch_size` really low).

So the solution has two elements:

- [x] species_lookup needs to be handled by a temporary table rather than a giant `CASE` expression (there's an alternative where it only combined formes and handles weird cases like "Mr. Mime" and then the reporting DAO sanitizes the rest, but that adds a lot of complexity)
- [ ] Inserts in the sinks need to done using the `.execute(insert(), values)` syntax rather than the current `.execute(insert().values(values)` syntax. AFAICT, this also necessitates storing the rows as `dict`s rather than tuples. This is, honestly, a better syntax, as it'll be clearer from the tests what's being tested than how it is now where I'm randomly selecting by index. But it's also going to be a PITA to rewrite all the tests.

--- a/.gitignore
+++ b/.gitignore
@@ -6,3 +6,4 @@
 sample_logs
 test.sqlite
 scratch.py
+.DS_Store

--- a/onix/backend/sql/dao.py
+++ b/onix/backend/sql/dao.py
@@ -214,7 +214,9 @@ class ReportingDAO(_dao.ReportingDAO):
                                       primary_key=True),
                             sa.Column('pretty', sa.String(64),
                                       nullable=False),
-                            prefixes=['TEMPORARY'])
+                            prefixes=['TEMPORARY'],
+                            keep_existing=True)
+
         sl_table.drop(bind=self.conn, checkfirst=True)
         sl_table.create(bind=self.conn)
 

--- a/onix/backend/sql/dao.py
+++ b/onix/backend/sql/dao.py
@@ -2,6 +2,7 @@
 import calendar
 import datetime
 
+from future.utils import iteritems
 import sqlalchemy as sa
 
 from onix import metrics
@@ -121,18 +122,15 @@ class ReportingDAO(_dao.ReportingDAO):
         result = self.conn.execute(query)
         return result.fetchone()[0] or 0
 
-    def _weighted_team_members(self, players, species_lookup):
+    def _weighted_team_members(self, players, sl_table):
         """
         Gets weights for individual team members and prettifies species names
 
         Args:
             players (sa.sql.expression.Alias) :
                 The relevant players with weights
-            species_lookup (dict) :
-                mapping of species names or forme-concatenations to their
-                display names. This is what handles things like determining
-                whether megas re tiered together or separately or what counts as
-                an "appearance-only" forme.
+            sl_table (sa.Table) :
+                table containing species mappings
 
         Returns:
             sa.sql.expression.Alias :
@@ -161,20 +159,28 @@ class ReportingDAO(_dao.ReportingDAO):
 
         combo_formes = sa.func.group_concat(join.c.species
                                             ).label('combined_formes')
-        pretty = sa.case(species_lookup,
-                         value=combo_formes,
-                         else_='-' + combo_formes)
 
-        query = (sa.select([join.c.bid,
-                            join.c.side,
-                            join.c.weight,
-                            join.c.slot,
-                            join.c.sid,
-                            pretty.label('species')])
-                 .select_from(join)
-                 .group_by(join.c.bid,
-                           join.c.side,
-                           join.c.slot))
+        by_combo_forme = (sa.select([join.c.bid,
+                                     join.c.side,
+                                     join.c.weight,
+                                     join.c.slot,
+                                     join.c.sid,
+                                     combo_formes])
+                          .select_from(join)
+                          .group_by(join.c.bid,
+                                    join.c.side,
+                                    join.c.slot)).alias()
+
+        join = by_combo_forme.join(sl_table, onclause=by_combo_forme.c.combined_formes == sl_table.c.species, isouter=True)
+
+        query = (sa.select([by_combo_forme.c.bid,
+                            by_combo_forme.c.side,
+                            by_combo_forme.c.weight,
+                            by_combo_forme.c.slot,
+                            by_combo_forme.c.sid,
+                            sa.func.ifnull(sl_table.c.pretty, '-' + by_combo_forme.c.combined_formes)
+                           .label('species')])
+                 .select_from(join))
         return query.alias()
 
     def _remove_duplicates(self, team_members):
@@ -202,12 +208,26 @@ class ReportingDAO(_dao.ReportingDAO):
 
     def get_usage_by_species(self, month, metagame, species_lookup,
                              baseline=1630.):
+
+        sl_table = sa.Table('species_lookup', schema.metadata,
+                            sa.Column('species', sa.String(512),
+                                      primary_key=True),
+                            sa.Column('pretty', sa.String(64),
+                                      nullable=False),
+                            prefixes=['TEMPORARY'])
+        sl_table.drop(bind=self.conn, checkfirst=True)
+        sl_table.create(bind=self.conn)
+
+        rows = [{'species': k, 'pretty': v} for k, v in
+                iteritems(species_lookup)]
+        self.conn.execute(sl_table.insert(), rows)
+
         team_members = self._remove_duplicates(
             self._weighted_team_members(
                 self._weighted_players(
                     self._filtered_battles(month, metagame),
                     baseline),
-                species_lookup))
+                sl_table))
 
         total = sa.func.sum(team_members.c.weight).label('sum')
         query = (sa.select([team_members.c.species,

--- a/onix/backend/sql/sinks.py
+++ b/onix/backend/sql/sinks.py
@@ -113,7 +113,7 @@ def convert_forme(forme):
     row = dict(id=compute_fid(forme),
                species=forme.species,
                ability=forme.ability)
-    row.update(forme.stats._asdict())
+    row.update(dict(zip(forme.stats._fields, forme.stats)))
     return row
 
 

--- a/onix/backend/sql/sinks.py
+++ b/onix/backend/sql/sinks.py
@@ -105,10 +105,16 @@ def convert_forme(forme):
         >>> forme = Forme('heatmor', 'gluttony',
         ...               PokeStats(333, 241, 170, 253, 150, 204))
         >>> row = convert_forme(forme)
-        >>> print(row) #doctest: +ELLIPSIS
-        ('379bf3...', 'heatmor', 'gluttony', 333, 241, 170, 253, 150, 204)
+        >>> print(row['ability'])
+        gluttony
+        >>> print(row['hp'])
+        333
     """
-    return (compute_fid(forme), forme.species, forme.ability) + forme.stats
+    row = dict(id=compute_fid(forme),
+               species=forme.species,
+               ability=forme.ability)
+    row.update(forme.stats._asdict())
+    return row
 
 
 def convert_moveset(sid, moveset):
@@ -135,27 +141,27 @@ def convert_moveset(sid, moveset):
         ...                   ['earthquake', 'rockslide', 'shadowclaw',
         ...                    'substitute'], 5, 255)
         >>> rows = convert_moveset('f4ce673a1', moveset)
-        >>> print(rows[schema.movesets]) #doctest: +ELLIPSIS
-        [('f4ce673...', 'm', 'leftovers', 5, 255)]
+        >>> print(rows[schema.movesets][0]['item'])
+        leftovers
         >>> print(sum(map(lambda x:len(x), rows.values())))
         7
     """
     rows = dict()
 
-    rows[schema.movesets] = [(sid,
-                              moveset.gender,
-                              moveset.item,
-                              moveset.level,
-                              moveset.happiness)]
+    rows[schema.movesets] = [dict(id=sid,
+                                  gender=moveset.gender,
+                                  item=moveset.item,
+                                  level=moveset.level,
+                                  happiness=moveset.happiness)]
 
-    rows[schema.moveslots] = [(sid, i, move)
+    rows[schema.moveslots] = [dict(sid=sid, idx=i, move=move)
                               for i, move in enumerate(moveset.moves)]
 
     formes = [convert_forme(forme) for forme in moveset.formes]
 
     rows[schema.formes] = formes
 
-    rows[schema.moveset_forme] = [(sid, forme[0], i == 0)
+    rows[schema.moveset_forme] = [dict(sid=sid, fid=forme['id'], prime=(i == 0))
                                   for i, forme in enumerate(formes)]
 
     return rows
@@ -177,16 +183,16 @@ def convert_team(team_sids):
     Examples:
         >>> from onix.backend.sql.sinks import convert_team
         >>> rows = convert_team(['ghi', 'abc', 'def'])
-        >>> print(rows[0][0]) #doctest +ELLIPSIS
+        >>> print(rows[0]['tid']) #doctest +ELLIPSIS
         8711a93...
-        >>> print(rows[1][2])
+        >>> print(rows[1]['sid'])
         def
 
     """
 
     team_sids.sort()
     tid = compute_tid(team_sids)
-    rows = [(tid, i, sid) for i, sid in enumerate(team_sids)]
+    rows = [dict(tid=tid, idx=i, sid=sid) for i, sid in enumerate(team_sids)]
     return rows
 
 
@@ -223,23 +229,25 @@ def convert_player(player, bid, side, tid):
         ...                                      'rd': 129.53915739500627,
         ...                                      'w': 40})
         >>> row = convert_player(player, 5134, 1, 'aac491ca1')
-        >>> print(row[4])
+        >>> print(row['w'])
         40
-        >>> print(row[6])
+        >>> print(row['t'])
         None
-        >>> print(row[10])
+        >>> print(row['rpr'])
         None
     """
-    ratings = tuple(player.rating.get(metric, None)
-                    for metric in ('w',
-                                   'l',
-                                   't',
-                                   'elo',
-                                   'r',
-                                   'rd',
-                                   'rpr',
-                                   'rprd'))
-    return (bid, side, player.id, tid) + ratings
+    ratings = {metric: player.rating.get(metric, None)
+               for metric in ('w',
+                              'l',
+                              't',
+                              'elo',
+                              'r',
+                              'rd',
+                              'rpr',
+                              'rprd')}
+    row = dict(bid=bid, side=side, pid=player.id, tid=tid)
+    row.update(ratings)
+    return row
 
 
 def convert_battle_info(battle_info):
@@ -266,11 +274,11 @@ def convert_battle_info(battle_info):
         ...                          [['abc', 'cab', 'bac'],
         ...                           ['123', '312', '213']], 16, 'forfeit')
         >>> rows = convert_battle_info(battle_info)
-        >>> print(rows[schema.battle_players][0][1])
+        >>> print(rows[schema.battle_players][0]['side'])
         1
-        >>> print(rows[schema.battle_players][0][3]) #doctest +ELLIPSIS
+        >>> print(rows[schema.battle_players][0]['tid']) #doctest +ELLIPSIS
         267e429f...
-        >>> print(rows[schema.teams][0][0])
+        >>> print(rows[schema.teams][0]['tid']) #doctest +ELLIPSIS
         267e429f...
     """
     rows = dict()
@@ -278,14 +286,14 @@ def convert_battle_info(battle_info):
     teams = [convert_team(team) for team in battle_info.slots]
     rows[schema.teams] = sum(teams, [])
 
-    rows[schema.battle_infos] = [(battle_info.id,
-                                  battle_info.format,
-                                  battle_info.date,
-                                  battle_info.turn_length,
-                                  battle_info.end_type)]
+    rows[schema.battle_infos] = [dict(id=battle_info.id,
+                                      format=battle_info.format,
+                                      date=battle_info.date,
+                                      turns=battle_info.turn_length,
+                                      end_type=battle_info.end_type)]
 
     rows[schema.battle_players] = [convert_player(player, battle_info.id,
-                                                  i + 1, teams[i][0][0])
+                                                  i + 1, teams[i][0]['tid'])
                                    for i, player
                                    in enumerate(battle_info.players)]
 
@@ -334,7 +342,7 @@ class _InsertHandler(object):
     def perform_inserts(self, connection):
         with connection.begin() as transaction:
             for table, rows in iteritems(self.cache):
-                connection.execute(table.insert().values(rows))
+                connection.execute(table.insert(), rows)
         self.cache = self._new_cache()
 
 

--- a/tests/backend/sql/test_sinks.py
+++ b/tests/backend/sql/test_sinks.py
@@ -116,9 +116,9 @@ def test_convert_forme():
                   PokeStats(349, 234, 208, 276, 239, 202))
     row = sinks.convert_forme(forme)
 
-    assert 'moltres' == row[1]
-    assert 'pressure' == row[2]
-    assert 208 == row[5]
+    assert 'moltres' == row['species']
+    assert 'pressure' == row['ability']
+    assert 208 == row['dfn']
 
 
 def test_convert_moveset():
@@ -134,12 +134,12 @@ def test_convert_moveset():
 
     rows = sinks.convert_moveset(expected_sid, moveset)
 
-    assert expected_sid == rows[schema.movesets][0][0]
-    assert 'tyranitarite' == rows[schema.movesets][0][2]
-    assert 'icepunch' == rows[schema.moveslots][2][2]
-    assert 187 == rows[schema.formes][1][8]
-    assert True == rows[schema.moveset_forme][0][2]
-    assert False == rows[schema.moveset_forme][1][2]
+    assert expected_sid == rows[schema.movesets][0]['id']
+    assert 'tyranitarite' == rows[schema.movesets][0]['item']
+    assert 'icepunch' == rows[schema.moveslots][2]['move']
+    assert 187 == rows[schema.formes][1]['spe']
+    assert rows[schema.moveset_forme][0]['prime']
+    assert not rows[schema.moveset_forme][1]['prime']
 
 
 def test_convert_team():
@@ -154,10 +154,10 @@ def test_convert_team():
 
     rows = sinks.convert_team(team_sids)
 
-    assert expected_tid == rows[0][0]
-    assert expected_tid == rows[3][0]
-    assert 4 == rows[4][1]
-    assert '0c913e97615a85c6' == rows[0][2]
+    assert expected_tid == rows[0]['tid']
+    assert expected_tid == rows[3]['tid']
+    assert 4 == rows[4]['idx']
+    assert '0c913e97615a85c6' == rows[0]['sid']
 
 
 def test_convert_player():
@@ -167,10 +167,10 @@ def test_convert_player():
                               'rd': None, 'w': 63})
 
     row = sinks.convert_player(player, 513591, 2, '5742af7b')
-    assert 'tazye' == row[2]
-    assert 2 == row[5]
-    assert row[9] is None
-    assert row[8] is None
+    assert 'tazye' == row['pid']
+    assert 2 == row['l']
+    assert row['rd'] is None
+    assert row['r'] is None
 
 
 def test_convert_battle_info():
@@ -183,12 +183,12 @@ def test_convert_battle_info():
 
     rows = sinks.convert_battle_info(battle_info)
 
-    assert 'anythinggoes' == rows[schema.battle_infos][0][1]
-    assert 1109.76 == rows[schema.battle_players][0][8]
-    assert 2 == rows[schema.battle_players][1][1]
-    assert rows[schema.teams][1][0] == rows[schema.teams][3][0]
-    assert rows[schema.teams][1][2] == rows[schema.teams][3][2]
-    assert rows[schema.teams][1][1] == rows[schema.teams][3][1]
+    assert 'anythinggoes' == rows[schema.battle_infos][0]['format']
+    assert 1109.76 == rows[schema.battle_players][0]['r']
+    assert 2 == rows[schema.battle_players][1]['side']
+    assert rows[schema.teams][1]['tid'] == rows[schema.teams][3]['tid']
+    assert rows[schema.teams][1]['sid'] == rows[schema.teams][3]['sid']
+    assert rows[schema.teams][1]['idx'] == rows[schema.teams][3]['idx']
 
 
 @pytest.mark.usefixtures('initialize_db')

--- a/tests/integration/test_pipeline.py
+++ b/tests/integration/test_pipeline.py
@@ -55,8 +55,8 @@ def log_folder(tmpdir, ctx):
              for setnums in ((0, 2, 3, 4, 5, 7),
                              (4, 6, 7, 8, 9, 10))]
 
-    players = [lg.generate_player('alice', rpr=3000.)[0],
-               lg.generate_player('bob', rpr=1630.)[0]]
+    players = [lg.generate_player('alice', rpr=3000., rprd=50.)[0],
+               lg.generate_player('bob', rpr=1630., rprd=50.)[0]]
 
     log = lg.generate_log(players, teams, 36, 'normal')
     json.dump(log, open(log_folder.mkdir('2016-09').mkdir('ou')

--- a/tests/integration/test_pipeline.py
+++ b/tests/integration/test_pipeline.py
@@ -1,4 +1,5 @@
 """Tests the entire Onix workflow"""
+from __future__ import division
 import json
 import random
 
@@ -23,20 +24,24 @@ def engine(tmpdir):
     create_tables(engine)
     return engine
 
+
 @pytest.fixture()
 def ctx():
     return contexts.get_standard_context()
 
 
 @pytest.fixture()
-def log_folder(tmpdir, ctx):
+def single_log(tmpdir, ctx):
     """
-    Generates logs and returns the location where the logs are stored
+    Generates a single log and returns a path to the log
 
     Returns:
-        str : the absolute path to the directory containing the stored logs
+        py.path.local : the absolute path to the generated log
     """
-    log_folder = tmpdir.mkdir('logs')
+    log_path = (tmpdir.mkdir('logs').mkdir('2016-09')
+                .mkdir('ou')
+                .mkdir('2016-09-30')
+                .join('battle-ou-134134.log.json'))
 
     movesets = [lg.generate_pokemon(species, ctx)[0]
                 for species in ('aerodactyl',  # 0
@@ -58,40 +63,114 @@ def log_folder(tmpdir, ctx):
     players = [lg.generate_player('alice', rpr=3000., rprd=50.)[0],
                lg.generate_player('bob', rpr=1630., rprd=50.)[0]]
 
-    log = lg.generate_log(players, teams, 36, 'normal')
-    json.dump(log, open(log_folder.mkdir('2016-09').mkdir('ou')
-                        .mkdir('2016-09-30')
-                        .join('battle-ou-134134.log.json').strpath, 'w+'))
+    json.dump(lg.generate_log(players, teams, 36, 'normal'),
+              open(log_path.strpath, 'w+'))
+
+    return log_path
+
+
+@pytest.fixture()
+def log_folder(tmpdir, ctx):
+    """
+    Generates a bunch of log and returns a path to the containing directory
+
+    Returns:
+        py.path.local : the absolute path to the log folder
+    """
+    log_folder = tmpdir.mkdir('logs')
+    meta_folder = log_folder.mkdir('2016-10').mkdir('ubers')
+
+    for i in range(1, 32):
+        meta_folder.mkdir('2016-10-{0:02}'.format(i))
+
+    pokemon = [('zygarde', 2),
+               ('yanmega', 4),
+               ('xerneas', 5),
+               ('wailord', 2),
+               ('volcarona', 6),
+               ('umbreon', 4),
+               ('tornadustherian', 5),
+               ('swampertmega', 2),
+               ('registeel', 3),
+               ('quagsire', 8),
+               ('pidgeotmega', 5),
+               ('octillery', 1),
+               ('nidoking', 3),
+               ('moltres', 7)]
+
+    movesets = [[lg.generate_pokemon(species, ctx)[0] for _ in range(num)]
+                for species, num in pokemon]
+
+    teams = [[random.choice(mon_sets)
+              for mon_sets in random.sample(movesets, 6)]
+             for _ in range(100)]
+
+    names = ('Alice', 'Bob', 'Charley', 'Delia', 'Eve', 'Frank', 'Geoff',
+             'Harold', 'Igor', 'Josephine')
+
+    end_types = ('normal', 'forfeit')
+
+    for i in range(1000):
+        log = lg.generate_log([lg.generate_player(random.choice(names))[0]
+                               for _ in range(2)],
+                              random.sample(teams, 2),
+                              random.randint(6, 128),
+                              random.choice(end_types))
+        path = meta_folder.join('2016-10-{0:02}'.format(
+            31 * i // 1000 + 1)).join('battle-ubers-{0}.log.json'.format(i))
+        json.dump(log, open(path.strpath, 'w+'))
 
     return log_folder
 
 
-def test_sql_backend_pipeline(engine, log_folder, ctx):
+class TestSqlBackendPipeline(object):
 
-    with engine.connect() as conn:
-        with sinks.MovesetSink(conn) as ms, sinks.BattleInfoSink(conn) as bs:
-            processor = LogProcessor(ms, bs, None)
+    def test_single_log(self, engine, single_log, ctx):
 
-            processor.process_logs(log_folder.strpath)
+        with engine.connect() as conn:
+            with sinks.MovesetSink(conn) as ms, sinks.BattleInfoSink(conn) as bs:
+                processor = LogProcessor(ms, bs, None)
 
-        rd = dao.ReportingDAO(conn)
+                processor.process_logs(single_log.strpath, ref_type='file')
 
-        expected = ' Total battles: 1\n'\
-                   ' Avg. weight / team: 0.750000\n'\
-                   ' + ---- + ------------------------- + --------- +\n'\
-                   ' | Rank | Species                   | Usage %   |\n'\
-                   ' + ---- + ------------------------- + --------- +\n'\
-                   ' |    1 | Dragalge                  | 100.0000% |\n'\
-                   ' |    2 | Espeon                    | 100.0000% |\n'\
-                   ' |    3 | Froslass                  | 100.0000% |\n'\
-                   ' |    4 | Aerodactyl                |  66.6667% |\n'\
-                   ' |    5 | Blissey                   |  66.6667% |\n'\
-                   ' |    6 | Clawitzer                 |  66.6667% |\n'\
-                   ' |    7 | Gigalith                  |  33.3333% |\n'\
-                   ' |    8 | Hydreigon                 |  33.3333% |\n'\
-                   ' |    9 | Illumise                  |  33.3333% |\n'\
-                   ' + ---- + ------------------------- + --------- +\n'
+        with engine.connect() as conn:
+            rd = dao.ReportingDAO(conn)
 
-        assert expected == generate_usage_stats(rd, ctx.species_lookup,
-                                                '201609', 'ou')
+            expected = ' Total battles: 1\n'\
+                       ' Avg. weight / team: 0.750000\n'\
+                       ' + ---- + ------------------------- + --------- +\n'\
+                       ' | Rank | Species                   | Usage %   |\n'\
+                       ' + ---- + ------------------------- + --------- +\n'\
+                       ' |    1 | Dragalge                  | 100.0000% |\n'\
+                       ' |    2 | Espeon                    | 100.0000% |\n'\
+                       ' |    3 | Froslass                  | 100.0000% |\n'\
+                       ' |    4 | Aerodactyl                |  66.6667% |\n'\
+                       ' |    5 | Blissey                   |  66.6667% |\n'\
+                       ' |    6 | Clawitzer                 |  66.6667% |\n'\
+                       ' |    7 | Gigalith                  |  33.3333% |\n'\
+                       ' |    8 | Hydreigon                 |  33.3333% |\n'\
+                       ' |    9 | Illumise                  |  33.3333% |\n'\
+                       ' + ---- + ------------------------- + --------- +\n'
+
+            assert expected == generate_usage_stats(rd, ctx.species_lookup,
+                                                    '201609', 'ou')
+
+    @pytest.mark.slow
+    def test_log_folder(self, engine, log_folder, ctx):
+        with engine.connect() as conn:
+            with sinks.MovesetSink(conn) as ms, sinks.BattleInfoSink(
+                    conn) as bs:
+                processor = LogProcessor(ms, bs, None)
+
+                processor.process_logs(log_folder.strpath)
+
+        with engine.connect() as conn:
+            rd = dao.ReportingDAO(conn)
+            report = generate_usage_stats(rd, ctx.species_lookup,
+                                          '201610', 'ubers', baseline=0)
+
+        lines = report.split('\n')
+        assert 21 == len(lines)  # 14 mons
+        assert ' Total battles: 1000' == lines[0]
+        assert ' Avg. weight / team: 1.000000' == lines[1]
 

